### DISCUSSION TITLE
Documentation/Impact of graph.set-vertex-id=true

### DISCUSSION THREAD

**Discussion by tungstn** [JanusGraph/janusgraph#2893] [Q&A] [answered]
Is there any additional documentation supporting the configuration option `graph.set-vertex-id=true`?

Here's the current docs from the [Configuration Reference](https://docs.janusgraph.org/v0.4/basics/configuration-reference/):
> Whether user provided vertex ids should be enabled and JanusGraph's automatic id allocation be disabled. Useful when operating JanusGraph in concert with another storage system that assigns long ids but disables some of JanusGraph's advanced features which can lead to inconsistent data. EXPERT FEATURE - USE WITH GREAT CARE.

The main phrase that I'm looking for clarification on is `disables some of JanusGraph's advanced features which can lead to inconsistent data`. 
* What are these `advanced features` (beyond vertex ID generation) which lead to `inconsistent data`?
* This flag seems to disable the auto-generation of vertex ids by janusgraph, and requires data to provide its own vertex id. Awesome - that's what I want!
* I can picture that if you use a poor methodology to create your vertex ids, it can lead to some bad situations... but that's outside of Janusgraph's concern IMO. (e.g., Conflicting IDs or IDs above the upper bounds of supported vertex IDs)
* But I don't follow what else within janusgraph itself may lead to inconsistent data if the only thing that changed are "who" creates vertex ids?
* Is there some other process that is impacted by this flag?

The verbiage clearly states this an expert feature and to use with caution, but I'm not sure I have enough context to evaluate our use case.

From initial, small-scale testing, this flag seems to function exactly like I think it does (just disable vertex id generation).

There are some SO/google articles with _related, but different_ questions, apologies if this has already been asked!

Thanks!

**Comment by li-boxuan** [comment, marked as answer]
That's a great question! This remark was from the original Titan authors (who are not active anymore), so you probably won't be able to get a "true" answer here. You might want to ask on the janusgraph-users mailing list and see others' use cases with this feature in production.

To my best knowledge with JanusGraph, I don't see why this feature would lead to inconsistent states, or how it would affect usage of so-called "advanced features".

**Reply by li-boxuan** [reply]
I just noticed two possible drawbacks of using custom id:

1. Obviously, you have to ensure your custom vertex ids are unique.
2. In JanusGraph, properties, and edges also have ids. These ids are auto-generated no matter if your vertex ids are auto-generated or not. There are a number of ID pools from which unique ids can be generated concurrently. A property/edge's id is generated from the same pool where its (incident) vertex is generated. If the vertex's id is user-provided, then it will always be the first ID pool that is used to generate IDs for its properties and edges. This might cause a load imbalance problem. I tend to believe this is a bug, but it might affect backward compatibility if we change this strategy (we could investigate it further, though).

**Reply by li-boxuan** [reply]
Btw FYI I am working on https://github.com/JanusGraph/janusgraph/issues/1221 which will support string custom Id. It will also overcome the second drawback I mentioned above.
